samples/python/onedim/diffusion_flame_continuation.py

```diff
@@ -54,43 +62,139 @@
+target_delta_Tmax = 20
+
+# Stop after this many successive errors
+max_error_count = 3
+error_count = 0
+
+# Stop after any failure if the strain rate has dropped to this fraction of the maximum
+strain_rate_tol = 0.10
+
 f.two_point_control_enabled = True
-spacing = 0.95
-temperature_increment = 10.0 # Kelvin
-maximum_temperature = []
-a_max = []
-for i in range(5000):
+f.max_time_step_count = 100
+Tmax = max(f.T)
+a_max = strain_rate = f.strain_rate('max')
+data = []  # integral output quantities for each step
+logger.info('Starting two-point control')
+
+for i in range(Nmax):
+    if strain_rate > 0.98 * a_max:
+        spacing = initial_spacing
+    else:
+        spacing = unstable_spacing
     control_temperature = np.min(f.T) + spacing*(np.max(f.T) - np.min(f.T))
-    print(f'Iteration {i}: Control temperature = {control_temperature} K')
+
+    # Store the flame state in case the iteration fails and we need to roll back
+    backup_state = f.to_array()
+
+    logger.debug(f'Iteration {i}: Control temperature = {control_temperature:.2f} K')
     f.set_left_control_point(control_temperature)
     f.set_right_control_point(control_temperature)
 
     # This decrement is what drives the two-point control. If failure
     # occurs, try decreasing the decrement.
     f.left_control_point_temperature -= temperature_increment
     f.right_control_point_temperature -= temperature_increment
+    f.clear_stats()
+
+    if (f.left_control_point_temperature < f.fuel_inlet.T + 100
+        or f.right_control_point_temperature < f.oxidizer_inlet.T + 100
+    ):
+        logger.info("Stopping because control point temperature is sufficiently close "
+                    "to inlet temperature.")
+        break
 
     try:
         f.solve(loglevel=0)
-    except ct.CanteraError as e:
-        print('Solver did not converge. Stopping.')
+        if abs(max(f.T) - Tmax) < 0.8 * target_delta_Tmax:
+            # Max temperature is changing slowly. Try a larger increment next step
+            temperature_increment = min(temperature_increment + 3, max_increment)
+        elif abs(max(f.T) - Tmax) > target_delta_Tmax:
+            # Max temperature is changing quickly. Scale down increment for next step
+            temperature_increment *= 0.9 * target_delta_Tmax / (abs(max(f.T) - Tmax))
+        error_count = 0
+    except ct.CanteraError as err:
+        logger.debug(err)
+        if strain_rate / a_max < strain_rate_tol:
+            logger.info('Traversed unstable branch down to '
+                f'{100 * strain_rate / a_max:.1f}% of the maximum strain rate.')
+            break
+
+        # Restore the previous solution and try a smaller temperature increment for the
+        # next iteration
+        f.from_array(backup_state)
+        temperature_increment = 0.7 * temperature_increment
+        error_count += 1
+        logger.warning(f"Solver did not converge on iteration {i}. Trying again with "
```

Comment:
The linter I am using (ruff) insists on lazy `%`-style formatting for logging strings.

Reply:
That seems like an unnecessary optimization here. I'd sooner switch back to simple `print` calls with f-strings than return to `%`-style string formatting.

Reply:
I honestly agree with you on that one.

Reply:
Unrelated to this PR, but the logger we use in the SCons code does support brace style formatting, that was an extension I think I added in the buildutils.

Reply:
Fwiw, I looked at this again, and it was actually Pylint ... [see link to W1203](https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/logging-fstring-interpolation.html), which is probably a little more aggressive than necessary.